Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h

```diff
@@ -77,20 +77,26 @@ class MANTID_GEOMETRY_DLL CSGObject : public IObject {
   /// Clone
   IObject *clone() const override { return new CSGObject(*this); }
 
+  IObject *cloneWithMaterial(const Kernel::Material &material) const override {
```

Comment:
The return should be **Covarient** otherwise you need dynamic_casts everywhere and the code becomes a total mess.